### PR TITLE
Fixing VeryRecursiveO with Conde

### LINKED ISSUE #2: very recursive breaks
```
(defrel (very-recursiveo)
  (conde
    ((nevero))
    ((very-recursiveo))
    ((alwayso))
    ((very-recursiveo))
    ((nevero))))
```

This causes an infinite loop and a stack overflow

```go
func VeryRecursiveO() micro.Goal {
	return Conde(
		micro.NeverO(),
		VeryRecursiveO(),
		micro.AlwaysO(),
		VeryRecursiveO(),
		micro.NeverO(),
	)
}
```

And this also creates a stack overflow, but in djsjointo

```go
func VeryRecursiveO() micro.Goal {
	return Conde(
		micro.NeverO(),
		func(s *micro.State) micro.StreamOfStates {
			return VeryRecursiveO()(s)
		},
		micro.AlwaysO(),
		func(s *micro.State) micro.StreamOfStates {
			return VeryRecursiveO()(s)
		},
		micro.NeverO(),
	)
}
```

Here is a test to run it
```go
func TestRecursive(t *testing.T) {
	micro.RunGoal(
		1,
		micro.CallFresh(func(q *ast.SExpr) micro.Goal {
			return VeryRecursiveO()
		}),
	)
}
```

And here is conde, which is probably incorrect
```go
/*
scheme code:

	(define-syntax conde
		(syntax-rules () (
			(conde (g ...) ...)
			(disj (conj g ...) ...)
		))
	)
*/
func Conde(gs ...micro.Goal) micro.Goal {
	if len(gs) == 0 {
		return micro.DisjointO(micro.ConjunctionO(gs...))
	}
	goals := make([]micro.Goal, len(gs))
	copy(goals, gs)
	goals[0] = micro.ConjunctionO(gs...)
	return micro.DisjointO(goals...)
}
```

### PR DESCRIPTION
**Pull request by deosjr**
Fixes #2

* Introduced the Zzz macro from minikanren to simplify delay.
* Added disj+ using Zzz
* Added a simplified version of Conde (real conde operates on list of
lists of goals)
* Caveat: two versions of Zzz, one of which also delaying the underlying
function call

### CODE REVIEW

File: mini/recursive.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+/*
+(defrel (very-recursiveo)
+  (conde
+    ((nevero))
+    ((very-recursiveo))
+    ((alwayso))
+    ((very-recursiveo))
+    ((nevero))))
+*/
+func VeryRecursiveO() micro.Goal {
```
Comment (awalterschulze) [inline mini/recursive.go:16]: Maybe this should move to the test file?

File: mini/conde.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		return micro.Mplus(g1s, g2s)
+	}
+}
+
+// TODO: should be list of lists of goals
+func Conde(gs ...micro.Goal) micro.Goal {
+	return DisjPlus(gs...)
+}
+
+func Zzz(g micro.Goal) micro.Goal {
```
Comment (awalterschulze) [inline mini/conde.go:34]: Oh I didn't know about disj+ or zzz, is this from the paper?
Maybe add a comment about how it should be used?

File: mini/conde.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	g2 := DisjPlus(gs[1:]...)
+	return func(s *micro.State) micro.StreamOfStates {
+		g1s := g1(s)
+		g2s := g2(s)
+		return micro.Mplus(g1s, g2s)
+	}
+}
+
+// TODO: should be list of lists of goals
+func Conde(gs ...micro.Goal) micro.Goal {
```
Comment (awalterschulze) [inline mini/conde.go:30]: Maybe just use disj+ until you figured out how to created a proper clone of conde?

**Review by awalterschulze — COMMENTED** [review]
It has been so long, I am already out of my depth reviewing this, so I just have nit picky comments.

**Comment by awalterschulze** [comment]
Don't worry about the travis diff error.
I will fix that at a later stage.

File: mini/recursive.go (hunk 1 above)
Comment (deosjr) [reply mini/recursive.go:16]: good idea

File: mini/conde.go (hunk 2 above)
Comment (deosjr) [reply mini/conde.go:34]: Yep both are under 5.1 Recovering miniKanren's control operators
I'll add a comment

File: mini/conde.go (hunk 3 above)
Comment (deosjr) [reply mini/conde.go:30]: I'll add the proper (full) definition of Conde, just wanted to keep it closer to your attempt in the issue first.
Will hurt readability a little

**Comment by deosjr** [comment]
Well, this just became quite the monster. Hope you still like it.
New version has the proper Zzz (which was Zzz2 before) which delays function calls.
The old Zzz is now DefRel, which is what is was actually doing anyways :)
Goals are fn() -> fn(state) -> stream

**Comment by awalterschulze** [comment]
Could GoalFn, also be Goalz? for a cuter name.  What do you think?

I think this looks great, really good job. Well done 👍

**Review by awalterschulze — APPROVED** [review]

**Pull request merged**